src/ray/core_worker/reference_count.cc

```diff
@@ -164,4 +315,339 @@ ReferenceCounter::GetAllReferenceCounts() const {
   return all_ref_counts;
 }
 
+void ReferenceCounter::GetAndClearLocalBorrowers(
+    const std::vector<ObjectID> &borrowed_ids,
+    ReferenceCounter::ReferenceTableProto *proto) {
+  absl::MutexLock lock(&mutex_);
+  ReferenceTable borrowed_refs;
+  for (const auto &borrowed_id : borrowed_ids) {
+    RAY_CHECK(GetAndClearLocalBorrowersInternal(borrowed_id, &borrowed_refs))
+        << borrowed_id;
+    // Decrease the ref count for each of the borrowed IDs. This is because we
+    // artificially increment each borrowed ID to keep it pinned during task
+    // execution. However, this should not count towards the final ref count
+    // returned to the task's caller.
+    auto it = borrowed_refs.find(borrowed_id);
+    if (it != borrowed_refs.end()) {
+      it->second.local_ref_count--;
+    }
+  }
+  ReferenceTableToProto(borrowed_refs, proto);
+}
+
+bool ReferenceCounter::GetAndClearLocalBorrowersInternal(const ObjectID &object_id,
+                                                         ReferenceTable *borrowed_refs) {
+  RAY_LOG(DEBUG) << "Pop " << object_id;
+  auto it = object_id_refs_.find(object_id);
+  if (it == object_id_refs_.end()) {
+    return false;
+  }
+
+  // We can only borrow objects that we do not own.
```

Comment:
How does this come up? If you submit tasks in a chain between two workers? Might be good to leave a comment as to why this isn't a `RAY_CHECK`